### PR TITLE
Schema changes for tags and using Akka Serialization for payloads

### PR DESCRIPTION
**Pull request by chbatey**
Raising so can keep track of progress. Plenty still to do

- [x] MySQL
- [x] Postgres
- [x] Oracle
- [x] H2
- [x] Metadata support
- [x] Decide what testing we want for the legacy dao
- [ ] Optional: reduce allocation for all the Option/Try wrangling

### CODE REVIEW

File: core/src/test/resources/schema/postgres/postgres-schema-v5.sql
Comment (patriknw) [inline core/src/test/resources/schema/postgres/postgres-schema-v5.sql:4]: alternative table name could be `event_journal` to pair with `event_tag`, but I don't have a strong opinion

**Review by patriknw — COMMENTED** [review]
looking good

File: core/src/test/resources/schema/postgres/postgres-schema-v5.sql
Comment (chbatey) [reply core/src/test/resources/schema/postgres/postgres-schema-v5.sql:12]: Yep, let's do that

File: core/src/test/resources/schema/postgres/postgres-schema-v5.sql
Comment (chbatey) [reply core/src/test/resources/schema/postgres/postgres-schema-v5.sql:5]: ah yes i'll add it anyway but as it is auto inserted it won't be null

File: core/src/test/resources/schema/postgres/postgres-schema-v5.sql
Comment (chbatey) [reply core/src/test/resources/schema/postgres/postgres-schema-v5.sql:4]: 👍

File: core/src/test/resources/schema/mysql/mysql-schema-v5.sql
Comment (chbatey) [reply core/src/test/resources/schema/mysql/mysql-schema-v5.sql]: Hadn't pushed mysql yet, now i have

File: core/src/main/scala/akka/persistence/jdbc/package.scala
Comment (octonato) [inline core/src/main/scala/akka/persistence/jdbc/package.scala:8]: can be removed
(also, last `c` was eaten)

**Review by octonato — COMMENTED** [review]
@chbatey, I did a first pass. I will do a more detailed one later. 
Left a few comments here and there. 

I think we should rebase this with #465 (if I get it green). Some important pieces have moved.

We should think how we want to test the old DAO, maybe we should have one job (Scala 2.13, JDK11 and old DAO).

One more thing, some headers are missing, this will show up in the build. Need to run a `sbt headersCreateAll`.

**Comment by octonato** [comment]
@chbatey, we merged two PRs (#465 and #474) today that will have considerable impact on this PR. I recommend you to rebase as soon as possible.

**Review by ennru — COMMENTED** [review]
I wonder if we can do better than "new" and "legacy" by calling them "v5" and "v4".

File: core/src/test/resources/oracle-application.conf
Comment (chbatey) [reply core/src/test/resources/oracle-application.conf:37]: 👍

File: core/src/test/scala/akka/persistence/jdbc/util/DropCreate.scala
Comment (chbatey) [reply core/src/test/scala/akka/persistence/jdbc/util/DropCreate.scala]: I am not so sure about the v4/v5 naming as it is v0->v4 and v5+. I'll let others chime in and we can go with the majority

File: core/src/main/scala/akka/persistence/jdbc/config/AkkaPersistenceConfig.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/jdbc/config/AkkaPersistenceConfig.scala:127]: We have a ticket for that. I think this PR is large enough without checking all the private/public access

File: core/src/main/scala/akka/persistence/jdbc/journal/dao/legacy/ByteArrayJournalDao.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/jdbc/journal/dao/legacy/ByteArrayJournalDao.scala:70]: We'd need other way to deal with sequence numbers e.g. a deleted_to table as well

File: core/src/main/scala/akka/persistence/jdbc/journal/dao/AkkaSerializerJournalDao.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/jdbc/journal/dao/AkkaSerializerJournalDao.scala:55]: Yep let's do that under a diff PR and have a deleted_to table if we think it is important enough.

It would be easy to make the change in a backward compatible way so no immediate need

File: core/src/test/resources/schema/postgres/postgres-schema-v5.sql
Comment (chbatey) [reply core/src/test/resources/schema/postgres/postgres-schema-v5.sql]: actually not needed for bigserial

**Comment by patriknw** [comment]
What is remaining here? Anything  that needs additional review?

**Comment by chbatey** [comment]
Ready for a final review. Nothing specific to look at.

File: core/src/main/scala/akka/persistence/jdbc/journal/dao/JournalQueries.scala
Comment (octonato) [inline core/src/main/scala/akka/persistence/jdbc/journal/dao/JournalQueries.scala:32]: here we could do: 

```scala
val (events, tags) = sorted.unzip
```

We would traverse it once instead of twice (2 maps) and we avoid using ._1 and ._2 (less readable).

File: core/src/main/scala/akka/persistence/jdbc/snapshot/dao/AkkaSerializerSnapshotDao.scala
Comment (octonato) [inline core/src/main/scala/akka/persistence/jdbc/snapshot/dao/AkkaSerializerSnapshotDao.scala:68]: this TODO is important to solve before we merge this one

I agree, we should not call `toOption`

File: core/src/main/scala/akka/persistence/jdbc/testkit/internal/SchemaType.scala
Comment (octonato) [inline core/src/main/scala/akka/persistence/jdbc/testkit/internal/SchemaType.scala:13]: why the extra `{}`?

File: core/src/main/scala/akka/persistence/jdbc/testkit/internal/SchemaUtilsImpl.scala
Comment (octonato) [inline core/src/main/scala/akka/persistence/jdbc/testkit/internal/SchemaUtilsImpl.scala:122]: I think we should invert this. Default one is the new and we add `legacy` or `v4` to previous one.

File: core/src/test/resources/oracle-schema-overrides.conf
Comment (octonato) [inline core/src/test/resources/oracle-schema-overrides.conf:4]: Oh, is that the reason why we are still supporting custom column names? How was this working before. 

I also find this strange, I don't think this is an issue in akka projections. At least not for oracle.

File: core/src/test/scala/akka/persistence/jdbc/query/JournalDaoStreamMessagesMemoryTest.scala
Comment (octonato) [inline core/src/test/scala/akka/persistence/jdbc/query/JournalDaoStreamMessagesMemoryTest.scala:58]: is that still true?

File: core/src/test/scala/akka/persistence/jdbc/query/JournalSequenceActorTest.scala
Comment (octonato) [inline core/src/test/scala/akka/persistence/jdbc/query/JournalSequenceActorTest.scala:44]: idem as above

**Review by octonato — APPROVED** [review]
Looking really good. 

I think we are almost there. I left a few comments and suggestions here and there. 


Some things needs to be tackled before merging tough.

File: core/src/main/resources/reference.conf
Comment (chbatey) [reply core/src/main/resources/reference.conf:156]: Yep it was in the end for Oracle

File: .travis.yml
Comment (chbatey) [reply .travis.yml:84]: Okay

File: core/src/test/resources/oracle-schema-overrides.conf
Comment (chbatey) [reply core/src/test/resources/oracle-schema-overrides.conf:4]: It is only an issue when you have a prepared statement that returns a value (in our case the ordering id), this isn't supported for case sensitive names in the oracle jdbc driver

File: core/src/main/resources/schema/mysql/mysql-create-schema-v5.sql
Comment (chbatey) [reply core/src/main/resources/schema/mysql/mysql-create-schema-v5.sql:24]: Sure

**Comment by chbatey** [comment]
> may I insist on changing the namesto DefaultSnapshotDao and DefaultReadJournalDao

If you must but i think it is inconsistent with the name of the other impl which does include how it serialises in the name

File: core/src/main/resources/schema/mysql/mysql-create-schema-v5.sql
Comment (chbatey) [reply core/src/main/resources/schema/mysql/mysql-create-schema-v5.sql:24]: However we don't need a unique id for each event, we have that with the persistence id and sequence nr, the only reason we have this is for ordering so maybe that is a better name

File: core/src/main/scala/akka/persistence/jdbc/testkit/internal/SchemaUtilsImpl.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/jdbc/testkit/internal/SchemaUtilsImpl.scala:122]: 👍

File: core/src/test/scala/akka/persistence/jdbc/query/JournalDaoStreamMessagesMemoryTest.scala
Comment (chbatey) [reply core/src/test/scala/akka/persistence/jdbc/query/JournalDaoStreamMessagesMemoryTest.scala:58]: Ya

Comment (chbatey) [reply core/src/test/scala/akka/persistence/jdbc/query/JournalDaoStreamMessagesMemoryTest.scala:58]: It interacts with the old schema. We could re-write it but it is tested with the old schema so not sure it is worth the effort

File: core/src/main/scala/akka/persistence/jdbc/testkit/internal/SchemaUtilsImpl.scala
Comment (ennru) [inline core/src/main/scala/akka/persistence/jdbc/testkit/internal/SchemaUtilsImpl.scala:43]: This test lives in multiple places, can it be centralized?

File: docs/src/main/paradox/index.md
Comment (ennru) [inline docs/src/main/paradox/index.md:66]: Link to the v5 SQLs.

File: core/src/main/scala/akka/persistence/jdbc/testkit/internal/SchemaUtilsImpl.scala
Comment (chbatey) [reply core/src/main/scala/akka/persistence/jdbc/testkit/internal/SchemaUtilsImpl.scala]: 👍

**Review by ennru — APPROVED** [review]
LGTM!

File: core/src/main/resources/schema/mysql/mysql-create-schema-v5.sql
Comment (octonato) [reply core/src/main/resources/schema/mysql/mysql-create-schema-v5.sql]: Correct, makes more sense now that you clarified it.

**Pull request merged**
